### PR TITLE
Tune smartcontract registration cache

### LINKED ISSUE #2417: is still in use?
https://github.com/AElfProject/AElf/blob/40ce8b375b50421ee20b59c003701f91098fa679/src/AElf.Kernel.SmartContract/Application/DeployContractAddressForkCacheHandler.cs#L8-L11

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in 40ce8b375b50421ee20b59c003701f91098fa679 when #2409 was merged. cc @AElfProject.

### LINKED ISSUE #2444: no fork check
https://github.com/AElfProject/AElf/blob/a5bc2e485ed926d696ea4a3014f4050c492434aa/src/AElf.Kernel.SmartContract/Application/IDeployedContractAddressProvider.cs#L14-L17

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in a5bc2e485ed926d696ea4a3014f4050c492434aa when #2426 was merged. cc @AElfProject.

### PR DESCRIPTION
**Pull request by jason-hoopox**
- [x] Tune smartcontract registration cache
- [x] Remove deployed contract address - fix #2417 , fix #2444

### CODE REVIEW

File: src/AElf.Kernel.Core/SmartContract/Application/IBlockchainStateService.cs [hunk 1]
```diff
@@ -87,5 +98,56 @@ public async Task RemoveBlockStateSetsAsync(IList<Hash> blockStateHashes)
...
         }
+        
+        public async Task<TEntity> GetBlockExecutedDataAsync<TEntity>(IChainContext chainContext)
+        {
+            var byteString = await _blockchainStateManager.GetStateAsync(typeof(TEntity).Name, chainContext.BlockHeight,
+                chainContext.BlockHash);
+            return SerializationHelper.Deserialize<TEntity>(byteString?.ToByteArray());
+        }
+
+        public async Task<TEntity> GetBlockExecutedDataAsync<TKey, TEntity>(IChainContext chainContext, TKey key)
```
Comment (loning) [inline src/AElf.Kernel.Core/SmartContract/Application/IBlockchainStateService.cs]: make this method as an extend method

File: src/AElf.Kernel.Core/SmartContract/Domain/IBlockchainStateManager.cs [hunk 2]
```diff
@@ -191,7 +192,7 @@ public async Task MergeBlockStateAsync(ChainStateInfo chainStateInfo, Hash block
         public async Task<ChainStateInfo> GetChainStateInfoAsync()
         {
             var o = await _chainStateInfoCollection.GetAsync(_chainId.ToStorageKey());
```
Comment (rosona) [inline src/AElf.Kernel.Core/SmartContract/Domain/IBlockchainStateManager.cs]: rename var o

File: src/AElf.Kernel.Core/SmartContract/Application/IBlockchainStateService.cs [hunk 3]
```diff
@@ -16,6 +18,15 @@ public interface IBlockchainStateService
...
         Task RemoveBlockStateSetsAsync(IList<Hash> blockStateHashes);
+        Task<TEntity> GetBlockExecutedDataAsync<TEntity>(IChainContext chainContext);
+
+        Task<TEntity> GetBlockExecutedDataAsync<TKey, TEntity>(IChainContext chainContext, TKey key);
+        
+        Task AddBlockExecutedDataAsync<TEntity>(Hash blockHash, TEntity blockExecutedData);
+        
+        Task AddBlockExecutedDataAsync<TKey, TEntity>(Hash blockHash, TKey key, TEntity value);
+        
+        Task AddBlockExecutedDataAsync<TKey, TEntity>(Hash blockHash, IDictionary<TKey, TEntity> blockExecutedData);
```
Comment (loning) [inline src/AElf.Kernel.Core/SmartContract/Application/IBlockchainStateService.cs]: only keep one Task AddBlockExecutedDataAsync<TEntity>(Hash blockHash, TEntity blockExecutedData);

the other make in extension method

**Comment by loning** [comment]
//TODO: 1. check in the pool. 
 //2. if not in the pool, get from SmartContractRegistrationCacheProvider.Get(chainContext,address) 
 //    SmartContractRegistrationCacheProvider.Get(chainContext,address) => BlockchainStateService.GetExecutedData(chainContext,key) 
 //And in my view, you can also implement a general MemoryCacheProvider for BlockchainStateService.GetExecutedData 

firstly, Please reimplement these

**Pull request merged**